piet/src/text.rs

```diff
@@ -41,10 +62,29 @@ pub trait TextLayoutBuilder {
 /// - If the text position is not at a code point or grapheme boundary, undesirable behavior may
 /// occur.
 ///
+// TODO update hit testing for line breaking
 pub trait TextLayout {
     /// Measure the advance width of the text.
     fn width(&self) -> f64;
 
+
+    /// Used for changing the width of a text layout. Given a width, returns a [`TextLayout`]
+    /// struct with recalculated lines and line metrics.
+    // TODO: Should this take &self or self? I'd say maybe; perhaps the old struct should be kept around
+    // to help with caching.
+    fn update_width(&self, new_width: f64) -> Self;
```

Comment:
Interesting question.

A lot comes down to how we expect this to work under the hood. Based on the name of the function, I would expect the signature,

```rust
fn update_width(&mut self, new_width: f64);
```

That is, it would mutate the existing `TextLayout` to reflect the new width.

If we think that the caller might want to keep around multiple `TextLayout`s, then I could imagine the signature you use here, but I would call it something like `new_layout_with_width`.

I would only take `self` by value if there was expensive internal state that we wanted to reuse between the old and new layouts, or if we wanted the API to enforce some contract whereby the previous layout is unusable after you have a new one.

Reply:
This is something I discussed with @raphlinus, having more of an immutable feel.

My main thought for using `self` was to enforce a contract, but I couldn't think of any.

So, I'll keep `&self` for now, but we'll see what the future holds.

Thanks for your thoughts, they're a great articulation of things I have a fuzzy grasp of but can't articulate very well yet.